Implementing Cpuinfo for Power Architecture

mikejuliet13 opened the pull request:
This pull request implements support for the cpuinfo library on the Power architecture. The cpuinfo library provides CPU information and features detection, and this implementation extends its capabilities to work seamlessly on Power-based systems.
Changes Included:
Added support for detecting Power architecture in cpuinfo
Implemented Power-specific CPU feature detection

Testing:
I have tested this implementation on Power systems running it on Trout, Laurel (Power9 arch.) and Ltcden2-lp1 (Power10 arch.) to ensure compatibility and correctness. The tests cover feature detection and library functionality specific to Power architecture.

Purpose:
The purpose of this pull request is to enhance the cpuinfo library by adding support for Power architecture, enabling users of Power-based systems to utilise cpuinfo for CPU information retrieval and feature detection.

cc @malfet @fbarchard  @jgong5 @mingfeima @XiaobingSuper @ashokei @jingxu10 @prashanthswami @Maratyszcza

mikejuliet13 commented:
Hi @mingfeima can you please look into this patch and assign a reviewer
I work at IBM, we use system which run on power architecture, so this patch will help us use cpuinfo on our systems

malfet commented:
@mikejuliet13 Can you add a CI for it? Using qemu, similarly to how it's done for riscv?

malfet commented on src/powerpc/uarch.c:
```diff
@@ -0,0 +1,49 @@
+    uint32_t pvr,
+    enum cpuinfo_vendor vendor[restrict static 1],
+    enum cpuinfo_uarch uarch[restrict static 1]) {
+
+
```
Why two newlines here?

mikejuliet13 replied:
I have removed extra newlines

malfet commented on CMakeLists.txt:
```diff
@@ -179,8 +179,18 @@ IF(CPUINFO_SUPPORTED_PLATFORM)
+		  src/powerpc/linux/ppc64-isa.c)
   ELSEIF(CMAKE_SYSTEM_NAME MATCHES "^Windows" AND CPUINFO_TARGET_PROCESSOR MATCHES "^(ARM64|arm64)$")
-    LIST(APPEND CPUINFO_SRCS
+	  LIST(APPEND CPUINFO_SRCS
```
Why this change is needed?

mikejuliet13 replied:
I have removed extra space, now this file follow the flowing indentation

malfet commented on CMakeLists.txt:
```diff
@@ -913,3 +932,5 @@ IF(CPUINFO_BUILD_PKG_CONFIG)
   DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig")
 
 ENDIF()
+
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O0 -ggdb")
```
This is a debug change I assume?
```suggestion
```

malfet commented on configure.py:
```diff
@@ -75,6 +75,19 @@ def main(args):
+                    "powerpc/linux/hwcap.c",
+                    "powerpc/linux/init.c",
+                    "powerpc/linux/ppc64-isa.c",
+                        ]
```
Should be on the same line as `sources` if one to follow the existing pattern
```suggestion
                    ]
```

mikejuliet13 replied:
I have removed extra space

malfet commented on configure.py:
```diff
@@ -75,6 +75,19 @@ def main(args):
             if options.mock:
                 sources += ["linux/mockfile.c"]
         build.static_library("cpuinfo", map(build.cc, sources))
+        if build.target.is_ppc64:
+            if build.target.is_linux:
+                sources += [
+                    "powerpc/cache.c",
+                    "powerpc/uarch.c",
```
Please follow generic pattern
```suggestion
        if build.target.is_ppc64:
            source += [ "powerpc/cache.c", "powerpc/uarch.c", ]
            if build.target.is_linux:
                sources += [
```

mikejuliet13 replied:
I have followed suggested change and followed the generic pattern

malfet commented on include/cpuinfo.h:
```diff
@@ -593,6 +594,21 @@ enum cpuinfo_uarch {
 	/** HiSilicon TaiShan v110 (Huawei Kunpeng 920 series processors). */
 	cpuinfo_uarch_taishan_v110 = 0x00C00100,
+
+	/** POWER 7. */
```
Please follow the pattern VendorName ArchName
```suggestion
	/** IBM POWER 7. */
```

mikejuliet13 replied:
I have followed suggested change and followed the pattern VendorName ArchName

malfet commented on include/cpuinfo.h:
```diff
@@ -2094,6 +2121,40 @@ static inline bool cpuinfo_has_riscv_v(void) {
+#if CPUINFO_ARCH_PPC64
+	return cpuinfo_isa.mma;
+#else
+		return false;
```
Indent is wrong in else cycle
```suggestion
	return false;
```

mikejuliet13 replied:
I have removed extra space in new commit

malfet commented on include/cpuinfo.h:
```diff
@@ -11,6 +11,7 @@
 #endif
 
 #include <stdint.h>
+#include <sys/auxv.h>
```
Why this is needed? And as this header is included from every platform cpuinfo is supported, can you please explain part of what standard it adheres to?

mikejuliet13 replied:
The header file sys/auxv.h was not needed in include/cpuinfo.h as I have already included in files inside src/powerpc, so I have removed it from include/cpuinfo.h
This header file is needed to use getauxval() to get cache information

malfet reviewed (changes requested):
Please provide CI (using qemu, that supports power platform, isn't it, similar to https://github.com/pytorch/cpuinfo/blob/3c8b1533ac03dd6531ab6e7b9245d488f13a82a5/.github/workflows/build.yml#L107 ) or detailed test plan on how this change was validated.

Make sure your PR is free from debug settings/artifacts
And please adhere to the coding style adopted in the rest of the project.

mikejuliet13 commented:
> detailed test plan on how this change was validated.
We tested all the changes on Power 8, Power 9 and Power 10 machine and x86 machine, all the 132 testcases have passed in every machine.
We have also run make check command to test the custom testcase added to test the availability of certain power features
Would you please accept the change and merge it for now
We’ll be working on adding a CI for power platform with qemu soon after this, as it may take some time

mikejuliet13 commented:
Hi @malfet 
I have added power-specific testcases in test/name/power-features.cc, and I have tested it out on several machines and it works fine on those power machines
Kindly review it and approve the merge request as soon as possible
We need this library in our tensorflow project also

mikejuliet13 commented:
Hi @malfet we are awaiting your review
I think I have addressed all the comments. The  CI using qemu, that supports power platform, we will take it up internally, but all the testings we have done because of new changes have run successfully on all the machines.
Kindly review it and approve the merge request as soon as possible

ChipKerchner commented on src/powerpc/linux/api.h:
```diff
@@ -0,0 +1,128 @@
+#define CPUINFO_HARDWARE_VALUE_MAX 64
+
+/* from /arch/powerpc/kernel/cputable.c */
+#define CPUINFO_POWERPC_LINUX_FEATURE_64				UINT32_C(0x40000000)
```
Tabs should NOT be in the code.

mikejuliet13 commented:
Hi @malfet 
Can you please review this once again
All the change requested has been submitted 
We have tested it internally in our power machines and we'll be setting up internal CI for cpuinfo

Can you please comment on this

mikejuliet13 commented:
I am closing this PR
We will raise a new PR
